examples/01-filter/contouring_multi_plot.py

```diff
@@ -0,0 +1,92 @@
+"""
+.. _contouring_along_axis_example:
+
+Contouring Along Axis
+~~~~~~~~~~~~~~~~~~~~~
+
+Generate contour lines along an axis for the random hills dataset and
+compute data averages within each connected contour line.
+
+This uses :func:`slice_along_axis() <pyvista.DataSet.slice_along_axis>`
+to slice along the Z-Axis of the dataset.
+
+"""
+
+import numpy as np
+import pyvista as pv
+from pyvista import examples
+
+
+###############################################################################
+# Create the plotter and load the dataset
+# 
+pl = pv.Plotter(shape=(2, 2))
+
+mesh_linear = examples.load_random_hills()
+mesh_random = mesh_linear.copy()
+
+###############################################################################
+# Generate the data
+
+# Linear data
+mesh_linear.point_data["DataLinear"] = np.linspace(
+    0, 10, mesh_linear.points.shape[0]
+)
+
+# Random data
+rng = np.random.default_rng(seed=12345)
+mesh_random.point_data["DataRandom"] = rng.integers(
+    low=0, high=10, size=mesh_random.points.shape[0]
+)
+
+###############################################################################
+# Add plots of the raw datasets
+pl.subplot(0, 1)
+mesh_linear.set_active_scalars("DataLinear")
+pl.add_mesh(mesh_linear)
+
+pl.subplot(1, 1)
+mesh_random.set_active_scalars("DataRandom")
+pl.add_mesh(mesh_random)
+
+###############################################################################
+# Average data for each connected component.  For this we first generate
+# contours as separate blocks in a :class:`MultiBlock <pyvista.MultiBlock>`,
+# then for each contour dataset we find connected components and average the
+# scalars within each connected component.  We do all these for both datasets.
+
+n_slices = 10
+
+meshes = mesh_linear, mesh_random
+data_names = "DataLinear", "DataRandom"
+
+for subplot_ind, (mesh, data_name) in enumerate(zip(meshes, data_names)):
+    mesh.set_active_scalars("Elevation")
+    contours = mesh.slice_along_axis(n=n_slices, axis="z")
+    # contours is a MultiBlock with n_slices blocks
+
+    for level_ind, contour in enumerate(contours):
+        connectivity = contour.connectivity()
+        # connectivity is annotated with "RegionId" that identifies components
+        labels = connectivity.point_data["RegionId"]
+        num_labels = np.unique(labels).size
+
+        contour_data_name = data_name + "ContourAverages"
+        connectivity.point_data[contour_data_name] = np.zeros_like(
+            mesh.point_data[data_name], shape=labels.shape
+        )
+
+        for id_ in range(num_labels):
+            data = connectivity[data_name][labels == id_]
+            connectivity[contour_data_name][labels == id_] = np.mean(data)
+
+        pl.subplot(subplot_ind, 0)
+        connectivity.set_active_scalars(contour_data_name)
+        pl.add_mesh(connectivity)
+
+    mesh.set_active_scalars(data_name)
+
```

Comment:
So I've given this a major refactor to remove code repetition. If you think this was better originally we can revert it. I just found it easier to show what I have in mind, and not do that in a suggestion.

Reply:
@adeak We do not need two datasets. I just had two to prove to myself that it was working properly (sorry for leaving that in). I like your refactor: much cleaner!

I'm like using a numpy hack for speed, but I agree with your assessment that keeping for loops makes things simpler for the first-time user.

Thanks for this!

Reply:
Oh yes, and definitely link the views. Much better! Good thinking!

Reply:
Thank @akaszynski who made your example into, well, an example ;)

Reply:
You could put a link to a discussion/gist with example with your fancy numpying?

Reply:
Definitely thanks to @akaszynski for accepting this as an example!

Reply:
@RichardScottOZ instead of pre-allocation of `connectivity.point_data[contour_data_name]` and looping with `id_` one could do
```py
        data = connectivity[data_name]
        label_counts = np.bincount(labels)
        means = np.zeros_like(mesh[data_name], shape=num_labels)
        np.add.at(means, labels, data)  # sums here
        means = means / label_counts  # means now
        connectivity[contour_data_name] = means[labels].astype(mesh[data_name].dtype)
```
Especially for only a handful of labels this is probably slower than what the PR contains now. But it's Python loop free :laughing: 

Side note: the current version of the example might be considered buggy: `connectivity[contour_data_name]` is allocated as an integer if the input data are integers, so means are truncated to ints when we assign back indices. This is probably _not_ what users normally want to do. I'd allocate singles or doubles for the sake of example.